Include/Rover/Range.hpp

```diff
@@ -0,0 +1,182 @@
+    \tparam Distribution The type of distribution to produce.
+    \tparam Engine The type of random engine to use.
+  */
+  template<typename Begin, typename End,
+    typename Distribution = default_distribution_t<
+      std::common_type_t<typename Begin::Type, typename End::Type>>,
+    typename Engine = default_engine_t<
+      std::common_type_t<typename Begin::Type, typename End::Type>>>
+  class Range : Noncopyable {
+    static_assert(is_reference_v<Begin>,
+        "The beginning of the range is not a Reference");
+    static_assert(is_reference_v<End>,
+        "The end of the range is not a Reference");
+
+    public:
+
+      using Type = std::common_type_t<typename Begin::Type, typename End::Type>;
+
+      //! Constructs a Range over an interval defined by its sub-generators.
+      /*!
+        \param begin The generator evaluating the beginning of the range.
+        \param end The generator evaluating the end of the range.
+        \param interval The type of interval.
+        \warning begin and end must define a valid non-empty interval. No check
+                 is made if an open interval is empty, which will result in an
+                 infinite loop in generate
+      */
+      template<typename BeginFwd, typename EndFwd>
+      Range(BeginFwd&& begin, EndFwd&& end, Interval interval =
+          Interval::CLOSED);
+
+      Range(Range&& other);
+
+      template<typename Session>
+      constexpr Type generate(Session& session);
+
+      auto build_session() const;
+
+    private:
+      Engine m_engine;
+      Begin m_begin;
+      End m_end;
+      Interval m_interval;
+  };
+
+  template<typename BeginFwd, typename EndFwd>
+  Range(BeginFwd&&, EndFwd&&, Interval = Interval::CLOSED) -> 
+      Range<ensure_reference_t<std::decay_t<BeginFwd>>,
+      ensure_reference_t<std::decay_t<EndFwd>>>;
+
+  template<typename Begin, typename End, typename Distribution, typename Engine>
+  template<typename BeginFwd, typename EndFwd>
+  Range<Begin, End, Distribution, Engine>::Range
+      (BeginFwd&& begin, EndFwd&& end, Interval interval)
+      : m_engine(std::random_device()()),
+        m_begin(std::forward<BeginFwd>(begin)),
+        m_end(std::forward<EndFwd>(end)),
+        m_interval(interval) {}
+
+  template<typename Begin, typename End, typename Distribution, typename Engine>
+  Range<Begin, End, Distribution, Engine>::Range
+      (Range<Begin, End, Distribution, Engine>&& other)
+      : m_begin(std::move(other.m_begin)),
+        m_end(std::move(other.m_end)),
+        m_interval(std::move(other.m_interval)) {}
+
+  template<typename Begin, typename End, typename Distribution, typename Engine>
+  template<typename Session>
+  constexpr typename Range<Begin, End, Distribution, Engine>::Type
+      Range<Begin, End, Distribution, Engine>::generate(Session& session) {
+    while(true) {
+      auto begin = m_begin.generate(session);
+      auto end = m_end.generate(session);
+      Details::check_range(begin, end);
```

Comment:
Since I can no longer ensure range validity, maybe it makes sense to make it more consistent by not checking begin < end either? It is like if one expects some sort of input error checking, she can mistakenly rely on that in other cases such as empty open range too.